pkg/controller/windowsmachineconfig/tracker.go

```diff
@@ -79,14 +75,110 @@ func newNodeRecord(ipAddress string, secretName string) nodeRecord {
 	}
 }
 
+// getWindowsVM returns a windowsVM interface to be used for running commands against
+func getWindowsVM(ipAddress, instanceID string, credentials Credentials) (types.WindowsVM, error) {
+	winVM := &types.Windows{}
+	windowsCredentials := types.NewCredentials(instanceID, ipAddress, credentials.Password, credentials.Username)
+	winVM.Credentials = windowsCredentials
+	// Set up Winrm client
+	if err := winVM.SetupWinRMClient(); err != nil {
+		return nil, errors.Wrap(err, "error instantiating winrm client")
+	}
+	// Set up SSH client
+	if err := winVM.GetSSHClient(); err != nil {
+		return nil, errors.Wrap(err, "error instantiating ssh client")
+	}
+	return winVM, nil
+}
+
+// getInstanceIP gets the instance IP address associated with a node
+func getInstanceIP(nodeList *v1.NodeList, instanceID string) (string, error) {
+	// Ignore the nodes that are not ready.
+	for _, node := range nodeList.Items {
+		for _, condition := range node.Status.Conditions {
+			if condition.Type != v1.NodeReady {
+				continue
+			}
+		}
+		if strings.Contains(node.Spec.ProviderID, instanceID) {
+			for _, address := range node.Status.Addresses {
+				// If external ip exists return it as it is used in AWS, if it doesn't return internal(in case of azure)
+				// TODO: collect all the ips and define a priority order, external then internal etc.
+				if address.Type == v1.NodeExternalIP && len(address.Address) > 0 {
+					return address.Address, nil
+				}
+			}
+		}
+	}
+	return "", errors.New("unable to find Windows Worker nodes")
+}
+
+// getCredsFromSecret gets the credentials associated with the instance.
+func getCredsFromSecret(k8sclientset *kubernetes.Clientset, instanceID string, operatorNS string) (Credentials, error) {
```

Comment:
Why not just return a `types.Credentials` instead of returning this intermediary object? It seems like its not used anywhere except to get a `types.Credentials`

Reply:
I don't have a preference. I did not the want the fn to grow big however I can change it.

Reply:
We should avoid intermediary objects.

Reply:
IMO that shouldn't be a reason to sacrifice the quality of code in the controller.

Reply:
Ya, just that it we're passing unnecessary argument like k8client to a function like `getWindowsVM` which makes it hard to move it to helper later also made me think it is worth having intermediate object. But I don't have a strong opinion one way or the other.

Reply:
I don't see `k8sClient` being passed to `getWindowsVM`. Which function are you talking about?

Reply:
If we go down the path of not having intermediate vars, we have to pass the k8sclient to `getWindows` to and call getCredentials within that.

Reply:
I am fine leaving it as it is for now and revisiting when we move to Machine API.